File: Windows/StreamVision/ViewModels/MainViewModel.cs
using System;
using System.Collections.ObjectModel;
using System.Linq;
using System.Threading.Tasks;
using System.Windows.Input;
using CommunityToolkit.Mvvm.ComponentModel;
using CommunityToolkit.Mvvm.Input;
using LibVLCSharp.Shared;
using StreamVision.Models;
using StreamVision.Services;

namespace StreamVision.ViewModels
{
    public partial class MainViewModel : ObservableObject, IDisposable
    {
        private readonly DatabaseService _databaseService;
        private readonly M3UParser _m3uParser;
        private readonly XtreamCodesService _xtreamService;
        private readonly EpgService _epgService;

        // LibVLCSharp pour la lecture vidéo
        private readonly LibVLC _libVLC;
        private LibVLCSharp.Shared.MediaPlayer? _mediaPlayer;

        public LibVLCSharp.Shared.MediaPlayer? MediaPlayer => _mediaPlayer;

        [ObservableProperty]
        private ObservableCollection<PlaylistSource> _playlistSources = new();

        [ObservableProperty]
        private ObservableCollection<ChannelGroup> _channelGroups = new();

        [ObservableProperty]
        private ObservableCollection<Channel> _filteredChannels = new();

        [ObservableProperty]
        private ObservableCollection<Channel> _favoriteChannels = new();

        [ObservableProperty]
        private ObservableCollection<Channel> _recentChannels = new();

        [ObservableProperty]
        private ObservableCollection<string> _categories = new();

        [ObservableProperty]
        private PlaylistSource? _selectedSource;

        // Computed properties for Netflix-style UI
        public bool HasRecentChannels => RecentChannels.Count > 0;
        public bool HasFavorites => FavoriteChannels.Count > 0;

        [ObservableProperty]
        private Channel? _currentChannel;

        [ObservableProperty]
        private EpgProgram? _currentProgram;

        [ObservableProperty]
        private string _searchQuery = string.Empty;

        [ObservableProperty]
        private bool _isLoading;

        [ObservableProperty]
        private string _statusMessage = "Ready";

        [ObservableProperty]
        private bool _isPlaying;

        [ObservableProperty]
        private bool _isMuted;

        [ObservableProperty]
        private int _volume = 100;

        [ObservableProperty]
        private bool _isFullScreen;

        [ObservableProperty]
        private string _currentView = "Channels";

        [ObservableProperty]
        private int _bufferSize = 2000;

        // Content type: Live, VOD, Series (comme Smarters Pro)
        [ObservableProperty]
        private string _contentType = "Live";

        [ObservableProperty]
        private ObservableCollection<MediaItem> _vodItems = new();

        [ObservableProperty]
        private ObservableCollection<MediaItem> _seriesItems = new();

        [ObservableProperty]
        private ObservableCollection<XtreamCategory> _currentCategories = new();

        public MainViewModel()
        {
            _databaseService = new DatabaseService();
            _m3uParser = new M3UParser();
            _xtreamService = new XtreamCodesService();
            _epgService = new EpgService();

            // Initialiser LibVLC
            Core.Initialize();
            _libVLC = new LibVLC("--network-caching=2000", "--live-caching=2000");
            _mediaPlayer = new LibVLCSharp.Shared.MediaPlayer(_libVLC);
            _mediaPlayer.Volume = _volume;
        }

        public async Task InitializeAsync()
        {
            try
            {
                IsLoading = true;
                StatusMessage = "Initializing...";

                // VLC initialization removed - video playback disabled

                // Initialize database
                await _databaseService.InitializeAsync();

                // Load data
                await LoadPlaylistSourcesAsync();
                await LoadFavoriteChannelsAsync();
                await LoadRecentChannelsAsync();

                StatusMessage = "Ready";
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error: {ex.Message}";
            }
            finally
            {
                IsLoading = false;
            }
        }

        private async Task LoadPlaylistSourcesAsync()
        {
            var sources = await _databaseService.GetPlaylistSourcesAsync();
            PlaylistSources.Clear();
            foreach (var source in sources)
            {
                PlaylistSources.Add(source);
            }

            if (PlaylistSources.Count > 0 && SelectedSource == null)
            {
                SelectedSource = PlaylistSources.First();
            }
        }

        private async Task LoadFavoriteChannelsAsync()
        {
            var favorites = await _databaseService.GetFavoriteChannelsAsync();
            FavoriteChannels.Clear();
            foreach (var channel in favorites)
            {
                FavoriteChannels.Add(channel);
            }
            OnPropertyChanged(nameof(HasFavorites));
        }

        private async Task LoadRecentChannelsAsync()
        {
            var recent = await _databaseService.GetRecentChannelsAsync();
            RecentChannels.Clear();
            foreach (var channel in recent)
            {
                RecentChannels.Add(channel);
            }
            OnPropertyChanged(nameof(HasRecentChannels));
        }

        partial void OnSelectedSourceChanged(PlaylistSource? value)
        {
            if (value != null)
            {
                _ = LoadChannelsForSourceAsync(value);
            }
        }

        private async Task LoadChannelsForSourceAsync(PlaylistSource source)
        {
            try
            {
                IsLoading = true;
                StatusMessage = $"Loading channels from {source.Name}...";

                System.Collections.Generic.List<Channel> channels;

                // Charger dynamiquement depuis l'API (comme Smarters Pro)
                if (source.Type == SourceType.XtreamCodes &&
                    !string.IsNullOrEmpty(source.Username) &&
                    !string.IsNullOrEmpty(source.Password))
                {
                    StatusMessage = "Connecting to Xtream Codes API...";
                    channels = await _xtreamService.GetLiveStreamsAsChannelsAsync(
                        source.Url, source.Username, source.Password, source.Id);
                }
                else
                {
                    // Fallback pour M3U (charger depuis cache DB)
                    channels = await _databaseService.GetChannelsAsync(source.Id);
                }

                OrganizeChannelsIntoGroups(channels);

                source.ChannelCount = channels.Count;
                StatusMessage = $"Loaded {channels.Count} channels";
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error: {ex.Message}";
            }
            finally
            {
                IsLoading = false;
            }
        }

        private void OrganizeChannelsIntoGroups(System.Collections.Generic.List<Channel> channels)
        {
            ChannelGroups.Clear();
            Categories.Clear();

            var grouped = channels.GroupBy(c => c.GroupTitle).OrderBy(g => g.Key);

            foreach (var group in grouped)
            {
                var channelGroup = new ChannelGroup { Name = group.Key };
                foreach (var channel in group.OrderBy(c => c.Order).ThenBy(c => c.Name))
                {
                    channelGroup.Channels.Add(channel);
                }
                ChannelGroups.Add(channelGroup);
                Categories.Add(group.Key);
            }

            FilterChannels();
        }

        public void FilterByCategory(string category)
        {
            SearchQuery = string.Empty;
            FilteredChannels.Clear();

            foreach (var group in ChannelGroups)
            {
                if (group.Name == category)
                {
                    foreach (var channel in group.Channels)
                    {
                        FilteredChannels.Add(channel);
                    }
                    break;
                }
            }

            OnPropertyChanged(nameof(FilteredChannels));
        }

        partial void OnSearchQueryChanged(string value)
        {
            FilterChannels();
        }

        private void FilterChannels()
        {
            FilteredChannels.Clear();

            foreach (var group in ChannelGroups)
            {
                foreach (var channel in group.Channels)
                {
                    if (string.IsNullOrWhiteSpace(SearchQuery) ||
                        channel.Name.Contains(SearchQuery, StringComparison.OrdinalIgnoreCase) ||
                        channel.GroupTitle.Contains(SearchQuery, StringComparison.OrdinalIgnoreCase))
                    {
                        FilteredChannels.Add(channel);
                    }
                }
            }
        }

        [RelayCommand]
        private async Task AddM3UPlaylistAsync(string url)
        {
            try
            {
                IsLoading = true;
                StatusMessage = "Adding M3U playlist...";

                var source = new PlaylistSource
                {
                    Name = "M3U Playlist",
                    Type = SourceType.M3U,
                    Url = url,
                    LastSync = DateTime.Now
                };

                var channels = await _m3uParser.ParseFromUrlAsync(url, source.Id);
                source.ChannelCount = channels.Count;

                await _databaseService.SavePlaylistSourceAsync(source);
                await _databaseService.SaveChannelsAsync(channels);

                PlaylistSources.Add(source);
                SelectedSource = source;

                StatusMessage = $"Added {channels.Count} channels";
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error: {ex.Message}";
            }
            finally
            {
                IsLoading = false;
            }
        }

        [RelayCommand]
        private async Task AddXtreamPlaylistAsync((string url, string username, string password) credentials)
        {
            try
            {
                IsLoading = true;
                StatusMessage = "Connecting to Xtream Codes...";

                var accountInfo = await _xtreamService.AuthenticateAsync(
                    credentials.url, credentials.username, credentials.password);

                if (accountInfo == null)
                {
                    StatusMessage = "Authentication failed - check credentials";
                    return;
                }

                // Stocker seulement les credentials (comme Smarters Pro)
                var source = new PlaylistSource
                {
                    Name = $"IPTV - {credentials.username}",
                    Type = SourceType.XtreamCodes,
                    Url = credentials.url,
                    Username = credentials.username,
                    Password = credentials.password,
                    LastSync = DateTime.Now
                };

                // Sauvegarder uniquement la source (pas les chaînes)
                await _databaseService.SavePlaylistSourceAsync(source);

                PlaylistSources.Add(source);
                SelectedSource = source; // Ceci déclenchera LoadChannelsForSourceAsync automatiquement

                StatusMessage = $"Account added - Status: {accountInfo.Status}";
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error: {ex.Message}";
            }
            finally
            {
                IsLoading = false;
            }
        }

        [RelayCommand]
        private async Task RefreshPlaylistAsync()
        {
            if (SelectedSource == null) return;

            try
            {
                IsLoading = true;
                StatusMessage = "Refreshing playlist...";

                await _databaseService.DeleteChannelsForSourceAsync(SelectedSource.Id);

                System.Collections.Generic.List<Channel> channels;
                if (SelectedSource.Type == SourceType.XtreamCodes)
                {
                    channels = await _xtreamService.GetLiveStreamsAsChannelsAsync(
                        SelectedSource.Url,
                        SelectedSource.Username!,
                        SelectedSource.Password!,
                        SelectedSource.Id);
                }
                else
                {
                    channels = await _m3uParser.ParseFromUrlAsync(SelectedSource.Url, SelectedSource.Id);
                }

                await _databaseService.SaveChannelsAsync(channels);
                SelectedSource.LastSync = DateTime.Now;
                SelectedSource.ChannelCount = channels.Count;
                await _databaseService.SavePlaylistSourceAsync(SelectedSource);

                await LoadChannelsForSourceAsync(SelectedSource);

                StatusMessage = $"Refreshed {channels.Count} channels";
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error: {ex.Message}";
            }
            finally
            {
                IsLoading = false;
            }
        }

        [RelayCommand]
        private async Task DeletePlaylistAsync(PlaylistSource source)
        {
            try
            {
                await _databaseService.DeletePlaylistSourceAsync(source.Id);
                PlaylistSources.Remove(source);

                if (SelectedSource == source)
                {
                    SelectedSource = PlaylistSources.FirstOrDefault();
                    if (SelectedSource == null)
                    {
                        ChannelGroups.Clear();
                        FilteredChannels.Clear();
                    }
                }

                StatusMessage = "Playlist deleted";
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error: {ex.Message}";
            }
        }

        [RelayCommand]
        private async Task PlayChannelAsync(Channel channel)
        {
            try
            {
                // Update current channel
                if (CurrentChannel != null)
                    CurrentChannel.IsPlaying = false;

                CurrentChannel = channel;
                channel.IsPlaying = true;

                // Lecture avec LibVLC
                if (_mediaPlayer != null && !string.IsNullOrEmpty(channel.StreamUrl))
                {
                    _mediaPlayer.Stop();
                    using var media = new Media(_libVLC, new Uri(channel.StreamUrl));
                    _mediaPlayer.Play(media);
                    IsPlaying = true;
                    StatusMessage = $"Playing: {channel.Name}";
                }

                // Update EPG
                if (!string.IsNullOrEmpty(channel.EpgId))
                {
                    CurrentProgram = _epgService.GetCurrentProgram(channel.EpgId);
                }

                // Add to recent
                await _databaseService.AddRecentChannelAsync(channel.Id);
                await LoadRecentChannelsAsync();
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error: {ex.Message}";
            }
        }

        [RelayCommand]
        private void PlayPause()
        {
            if (_mediaPlayer == null) return;

            if (_mediaPlayer.IsPlaying)
            {
                _mediaPlayer.Pause();
                IsPlaying = false;
            }
            else
            {
                _mediaPlayer.Play();
                IsPlaying = true;
            }
        }

        [RelayCommand]
        private void Stop()
        {
            _mediaPlayer?.Stop();
            if (CurrentChannel != null)
                CurrentChannel.IsPlaying = false;
            CurrentChannel = null;
            CurrentProgram = null;
            IsPlaying = false;
            StatusMessage = "Stopped";
        }

        [RelayCommand]
        private void ToggleMute()
        {
            if (_mediaPlayer == null) return;
            IsMuted = !IsMuted;
            _mediaPlayer.Mute = IsMuted;
        }

        partial void OnVolumeChanged(int value)
        {
            if (_mediaPlayer != null)
                _mediaPlayer.Volume = value;
        }

        [RelayCommand]
        private async Task ToggleFavoriteAsync(Channel channel)
        {
            channel.IsFavorite = !channel.IsFavorite;
            await _databaseService.UpdateChannelFavoriteAsync(channel.Id, channel.IsFavorite);
            await LoadFavoriteChannelsAsync();
        }

        [RelayCommand]
        private async Task LoadEpgAsync(string epgUrl)
        {
            try
            {
                IsLoading = true;
                StatusMessage = "Loading EPG...";

                await _epgService.LoadEpgFromUrlAsync(epgUrl);

                if (SelectedSource != null)
                {
                    SelectedSource.EpgUrl = epgUrl;
                    await _databaseService.SavePlaylistSourceAsync(SelectedSource);
                }

                StatusMessage = "EPG loaded successfully";
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error loading EPG: {ex.Message}";
            }
            finally
            {
                IsLoading = false;
            }
        }

        [RelayCommand]
        private void ChangeView(string view)
        {
            CurrentView = view;
        }

        [RelayCommand]
        private async Task SwitchContentTypeAsync(string contentType)
        {
            if (SelectedSource == null || SelectedSource.Type != SourceType.XtreamCodes) return;

            ContentType = contentType;

            try
            {
                IsLoading = true;
                StatusMessage = $"Loading {contentType}...";

                switch (contentType)
                {
                    case "Live":
                        await LoadChannelsForSourceAsync(SelectedSource);
                        break;
                    case "VOD":
                        await LoadVodContentAsync();
                        break;
                    case "Series":
                        await LoadSeriesContentAsync();
                        break;
                }
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error: {ex.Message}";
            }
            finally
            {
                IsLoading = false;
            }
        }

        private async Task LoadVodContentAsync()
        {
            if (SelectedSource == null) return;

            var items = await _xtreamService.GetVodStreamsAsync(
                SelectedSource.Url,
                SelectedSource.Username!,
                SelectedSource.Password!,
                SelectedSource.Id);

            VodItems.Clear();
            foreach (var item in items)
            {
                VodItems.Add(item);
            }

            // Organiser par catégorie
            var categories = await _xtreamService.GetVodCategoriesAsync(
                SelectedSource.Url,
                SelectedSource.Username!,
                SelectedSource.Password!);

            CurrentCategories.Clear();
            foreach (var cat in categories)
            {
                CurrentCategories.Add(cat);
            }

            StatusMessage = $"Loaded {items.Count} movies";
        }

        private async Task LoadSeriesContentAsync()
        {
            if (SelectedSource == null) return;

            var items = await _xtreamService.GetSeriesAsync(
                SelectedSource.Url,
                SelectedSource.Username!,
                SelectedSource.Password!,
                SelectedSource.Id);

            SeriesItems.Clear();
            foreach (var item in items)
            {
                SeriesItems.Add(item);
            }

            // Organiser par catégorie
            var categories = await _xtreamService.GetSeriesCategoriesAsync(
                SelectedSource.Url,
                SelectedSource.Username!,
                SelectedSource.Password!);

            CurrentCategories.Clear();
            foreach (var cat in categories)
            {
                CurrentCategories.Add(cat);
            }

            StatusMessage = $"Loaded {items.Count} series";
        }

        [RelayCommand]
        private async Task LoadSeriesEpisodesAsync(string seriesId)
        {
            if (SelectedSource == null) return;

            try
            {
                IsLoading = true;
                StatusMessage = "Loading episodes...";

                var seriesInfo = await _xtreamService.GetSeriesInfoAsync(
                    SelectedSource.Url,
                    SelectedSource.Username!,
                    SelectedSource.Password!,
                    seriesId);

                if (seriesInfo != null)
                {
                    var totalEpisodes = seriesInfo.Seasons.Sum(s => s.Episodes.Count);
                    StatusMessage = $"Loaded {seriesInfo.Seasons.Count} seasons, {totalEpisodes} episodes";
                    // TODO: Display in UI
                }
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error: {ex.Message}";
            }
            finally
            {
                IsLoading = false;
            }
        }

        [RelayCommand]
        private async Task PlayMediaItemAsync(MediaItem item)
        {
            try
            {
                StatusMessage = $"Playing: {item.Name}";
                // TODO: Integrate video player
                await Task.CompletedTask;
            }
            catch (Exception ex)
            {
                StatusMessage = $"Error: {ex.Message}";
            }
        }

        [RelayCommand]
        private void ToggleFullScreen()
        {
            IsFullScreen = !IsFullScreen;
        }

        [RelayCommand]
        private void PreviousChannel()
        {
            if (CurrentChannel == null) return;

            var currentIndex = FilteredChannels.IndexOf(CurrentChannel);
            if (currentIndex > 0)
            {
                PlayChannelCommand.Execute(FilteredChannels[currentIndex - 1]);
            }
        }

        [RelayCommand]
        private void NextChannel()
        {
            if (CurrentChannel == null) return;

            var currentIndex = FilteredChannels.IndexOf(CurrentChannel);
            if (currentIndex < FilteredChannels.Count - 1)
            {
                PlayChannelCommand.Execute(FilteredChannels[currentIndex + 1]);
            }
        }

        public void Dispose()
        {
            _mediaPlayer?.Stop();
            _mediaPlayer?.Dispose();
            _libVLC?.Dispose();
            _databaseService.Dispose();
        }
    }
}

File: macOS/StreamVision/Sources/ViewModels/MainViewModel.swift
import Foundation
import AVFoundation
import Combine

@MainActor
class MainViewModel: ObservableObject {

    // MARK: - Published Properties

    @Published var playlistSources: [PlaylistSource] = []
    @Published var selectedSource: PlaylistSource?
    @Published var channelGroups: [ChannelGroup] = []
    @Published var filteredChannels: [Channel] = []
    @Published var favoriteChannels: [Channel] = []
    @Published var recentChannels: [Channel] = []

    @Published var currentChannel: Channel?
    @Published var currentProgram: EpgProgram?

    @Published var searchQuery: String = "" {
        didSet { filterChannels() }
    }

    @Published var isLoading: Bool = false
    @Published var statusMessage: String = "Ready"

    @Published var isPlaying: Bool = false
    @Published var isMuted: Bool = false
    @Published var volume: Float = 1.0 {
        didSet { player?.volume = volume }
    }

    @Published var currentView: NavigationView = .channels
    @Published var showAddPlaylistSheet: Bool = false

    // MARK: - Player

    private(set) var player: AVPlayer?
    private var playerObserver: Any?

    // MARK: - Services

    private let dataStore = DataStore()
    private let m3uParser = M3UParser()
    private let xtreamService = XtreamCodesService()

    private var allChannels: [Channel] = []
    private var favoriteIds: Set<UUID> = []

    // MARK: - Navigation

    enum NavigationView {
        case channels
        case favorites
        case recent
        case epg
    }

    // MARK: - Initialization

    init() {
        player = AVPlayer()
        setupPlayerObserver()
    }

    private func setupPlayerObserver() {
        playerObserver = player?.observe(\.timeControlStatus, options: [.new]) { [weak self] player, _ in
            Task { @MainActor in
                self?.isPlaying = player.timeControlStatus == .playing
            }
        }
    }

    // MARK: - Loading Data

    func loadData() async {
        isLoading = true
        statusMessage = "Loading..."

        do {
            playlistSources = try await dataStore.loadPlaylistSources()
            favoriteIds = try await dataStore.loadFavorites()

            if let firstSource = playlistSources.first {
                selectedSource = firstSource
                await loadChannelsForSource(firstSource)
            }

            await loadFavoriteChannels()
            await loadRecentChannels()

            statusMessage = "Ready"
        } catch {
            statusMessage = "Error loading data: \(error.localizedDescription)"
        }

        isLoading = false
    }

    func loadChannelsForSource(_ source: PlaylistSource) async {
        isLoading = true
        statusMessage = "Loading channels..."

        do {
            allChannels = try await dataStore.loadChannels(forSourceId: source.id)

            // Update favorite status
            for i in allChannels.indices {
                allChannels[i].isFavorite = favoriteIds.contains(allChannels[i].id)
            }

            organizeChannelsIntoGroups()
            statusMessage = "Loaded \(allChannels.count) channels"
        } catch {
            statusMessage = "Error: \(error.localizedDescription)"
        }

        isLoading = false
    }

    private func organizeChannelsIntoGroups() {
        let grouped = Dictionary(grouping: allChannels) { $0.groupTitle }
        channelGroups = grouped.map { ChannelGroup(name: $0.key, channels: $0.value.sorted { $0.order < $1.order }) }
            .sorted { $0.name < $1.name }
        filterChannels()
    }

    private func filterChannels() {
        if searchQuery.isEmpty {
            filteredChannels = allChannels
        } else {
            filteredChannels = allChannels.filter {
                $0.name.localizedCaseInsensitiveContains(searchQuery) ||
                $0.groupTitle.localizedCaseInsensitiveContains(searchQuery)
            }
        }
    }

    private func loadFavoriteChannels() async {
        favoriteChannels = allChannels.filter { favoriteIds.contains($0.id) }
    }

    private func loadRecentChannels() async {
        do {
            let recentIds = try await dataStore.loadRecentChannels()
            recentChannels = recentIds.compactMap { id in
                allChannels.first { $0.id == id }
            }
        } catch {
            recentChannels = []
        }
    }

    // MARK: - Playlist Management

    func addM3UPlaylist(name: String, url: String, epgUrl: String?) async {
        isLoading = true
        statusMessage = "Adding M3U playlist..."

        do {
            guard let playlistURL = URL(string: url) else {
                statusMessage = "Invalid URL"
                isLoading = false
                return
            }

            var source = PlaylistSource(
                name: name.isEmpty ? "M3U Playlist" : name,
                type: .m3u,
                url: url,
                epgUrl: epgUrl
            )

            let channels = try await m3uParser.parseFromURL(playlistURL, sourceId: source.id)
            source.channelCount = channels.count

            try await dataStore.saveChannels(channels, forSourceId: source.id)

            playlistSources.append(source)
            try await dataStore.savePlaylistSources(playlistSources)

            selectedSource = source
            await loadChannelsForSource(source)

            statusMessage = "Added \(channels.count) channels"
        } catch {
            statusMessage = "Error: \(error.localizedDescription)"
        }

        isLoading = false
    }

    func addXtreamPlaylist(name: String, serverUrl: String, username: String, password: String) async {
        isLoading = true
        statusMessage = "Connecting to Xtream Codes..."

        do {
            guard let _ = try await xtreamService.authenticate(serverUrl: serverUrl, username: username, password: password) else {
                statusMessage = "Authentication failed"
                isLoading = false
                return
            }

            var source = PlaylistSource(
                name: name.isEmpty ? "Xtream - \(username)" : name,
                type: .xtreamCodes,
                url: serverUrl,
                username: username,
                password: password
            )

            statusMessage = "Loading channels..."
            let channels = try await xtreamService.getLiveStreams(
                serverUrl: serverUrl,
                username: username,
                password: password,
                sourceId: source.id
            )
            source.channelCount = channels.count

            try await dataStore.saveChannels(channels, forSourceId: source.id)

            playlistSources.append(source)
            try await dataStore.savePlaylistSources(playlistSources)

            selectedSource = source
            await loadChannelsForSource(source)

            statusMessage = "Added \(channels.count) channels"
        } catch {
            statusMessage = "Error: \(error.localizedDescription)"
        }

        isLoading = false
    }

    func refreshPlaylist() async {
        guard let source = selectedSource else { return }

        isLoading = true
        statusMessage = "Refreshing playlist..."

        do {
            try await dataStore.deleteChannels(forSourceId: source.id)

            var channels: [Channel]
            if source.type == .xtreamCodes, let username = source.username, let password = source.password {
                channels = try await xtreamService.getLiveStreams(
                    serverUrl: source.url,
                    username: username,
                    password: password,
                    sourceId: source.id
                )
            } else {
                guard let url = URL(string: source.url) else {
                    statusMessage = "Invalid URL"
                    isLoading = false
                    return
                }
                channels = try await m3uParser.parseFromURL(url, sourceId: source.id)
            }

            try await dataStore.saveChannels(channels, forSourceId: source.id)

            // Update source
            if let index = playlistSources.firstIndex(where: { $0.id == source.id }) {
                playlistSources[index].lastSync = Date()
                playlistSources[index].channelCount = channels.count
                selectedSource = playlistSources[index]
                try await dataStore.savePlaylistSources(playlistSources)
            }

            await loadChannelsForSource(source)
            statusMessage = "Refreshed \(channels.count) channels"
        } catch {
            statusMessage = "Error: \(error.localizedDescription)"
        }

        isLoading = false
    }

    func deletePlaylist(_ source: PlaylistSource) async {
        do {
            try await dataStore.deleteChannels(forSourceId: source.id)
            playlistSources.removeAll { $0.id == source.id }
            try await dataStore.savePlaylistSources(playlistSources)

            if selectedSource?.id == source.id {
                selectedSource = playlistSources.first
                if let newSource = selectedSource {
                    await loadChannelsForSource(newSource)
                } else {
                    channelGroups = []
                    filteredChannels = []
                }
            }

            statusMessage = "Playlist deleted"
        } catch {
            statusMessage = "Error: \(error.localizedDescription)"
        }
    }

    // MARK: - Playback

    func playChannel(_ channel: Channel) {
        currentChannel?.isFavorite = favoriteIds.contains(currentChannel?.id ?? UUID())

        currentChannel = channel

        guard let url = URL(string: channel.streamUrl) else {
            statusMessage = "Invalid stream URL"
            return
        }

        let playerItem = AVPlayerItem(url: url)
        player?.replaceCurrentItem(with: playerItem)
        player?.play()

        statusMessage = "Playing: \(channel.name)"

        // Add to recent
        Task {
            try? await dataStore.addRecentChannel(channel.id)
            await loadRecentChannels()
        }
    }

    func togglePlayPause() {
        if isPlaying {
            player?.pause()
        } else {
            player?.play()
        }
    }

    func stop() {
        player?.pause()
        player?.replaceCurrentItem(with: nil)
        currentChannel = nil
        currentProgram = nil
        statusMessage = "Stopped"
    }

    func toggleMute() {
        isMuted.toggle()
        player?.isMuted = isMuted
    }

    func previousChannel() {
        guard let current = currentChannel,
              let index = filteredChannels.firstIndex(of: current),
              index > 0 else { return }
        playChannel(filteredChannels[index - 1])
    }

    func nextChannel() {
        guard let current = currentChannel,
              let index = filteredChannels.firstIndex(of: current),
              index < filteredChannels.count - 1 else { return }
        playChannel(filteredChannels[index + 1])
    }

    // MARK: - Favorites

    func toggleFavorite(_ channel: Channel) async {
        if favoriteIds.contains(channel.id) {
            favoriteIds.remove(channel.id)
        } else {
            favoriteIds.insert(channel.id)
        }

        // Update channel in list
        if let index = allChannels.firstIndex(where: { $0.id == channel.id }) {
            allChannels[index].isFavorite = favoriteIds.contains(channel.id)
        }

        do {
            try await dataStore.saveFavorites(favoriteIds)
            await loadFavoriteChannels()
            filterChannels()
        } catch {
            statusMessage = "Error saving favorites"
        }
    }
}
